### ISSUE TITLE
[MNG-6511] Option -pl ! foo should not fail if foo does not exist

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7349]
**[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)** opened **[MNG-6511](https://issues.apache.org/jira/browse/MNG-6511?redirect=false)** and commented

While I completely understand why Maven throws an error when `\-pl/--projects` defines/contains a non-existing project, I don't really see why the negation of a non-existing project yields the same error, e.g.:

```
c:\_dev\git\gitflow-incremental-builder>mvn -pl !foo
[INFO] Scanning for projects...
[ERROR] [ERROR] Could not find the selected project in the reactor: foo @
[ERROR] Could not find the selected project in the reactor: foo -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MavenExecutionException
```

I'd say that at most this should be a warning, not an error.

This change would come in handy to reuse scripts with certain default options (e.g. quickly build everything without tests, checkstyle, _exclude moduleX_, etc.) on different hierarchy levels of larger multi module project.

---

**Affects:** 3.3.9, 3.6.0

**Issue Links:**
- [MNG-7033](https://issues.apache.org/jira/browse/MNG-7033) Excluded projects through -pl should not need to be in the reactor
 (_**"is duplicated by"**_)
- [MNG-7051](https://issues.apache.org/jira/browse/MNG-7051) Optionally skip non-existing profiles

- [MNG-6012](https://issues.apache.org/jira/browse/MNG-6012) Missing profile is only notified at the end of a run
 (_**"is required by"**_)

**Remote Links:**
- [GitHub Pull Request #100
](https://github.com/apache/maven-integration-testing/pull/100)
- [GitHub Pull Request #446
](https://github.com/apache/maven/pull/446)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1)

1 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)** commented

> In general that means strict is better to match the expected result.

I'd still say that a user first and foremost just expects the given module not to be built.
Anyhow, we have different opinons and it's of course up to you to decide (otherwise).

> Since it seems you're already using scripts, maybe mavenrc is an option or the Maven Project options introduced with MNG-5767 which seems to be porely documented.

You can't put `-pl ! ...` in `.mvn/maven.config` because it kicks in regardless of which level you are operating on in your multi module project. `mvn` will thus fail if you change the working directory or use `-f` in such a way that one of the negated modules is not part of the reactor anymore.
I guess the same happens with mavenrc.
Anyway, I would _not_ put `-pl ! ...` in either of these files because that would be like a permanent filter for the entire project causing those modules never to be built.
Instead, I would put `-pl ! ...` into my "quick build" script (and maybe other variants of that script) which lives happily next to my "pre push" script which doesn't exclude anything (or just use `mvn clean install` or whatever).

For me the bottom line of this discussion is that the current behaviour won't be changed, unfortunately.
I see one possible extension of `-pl`, though:
What if you could define something like `-pl !?foo`, meaning "exclude foo, if exists"? And Wildcard or even Regex support would be even more flexible...
WDYT?

**Comment by jira-importer** [comment]
**[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)** commented

MNG-7051 added `?` to profiles, so it now makes even more sense to have it for `-pl`.

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)

> and would even push it a little bit further: -pl ?bar should also be possible.

What do you expect to happen in case there is no such project? An INFO printout and that's it?

**Comment by jira-importer** [comment]
**[Martin Kanters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=martinkanters)** commented

[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)
[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders) and I are working on this. PRs are created already, but there is still some work to be done by us. I expect that to be resolved somewhere end of this week, early next week
core: https://github.com/apache/maven/pull/446
ITs: https://github.com/apache/maven-integration-testing/pull/100

If the optional project does not exist, it will log a warning (consistent with profiles).

**Comment by jira-importer** [comment]
**[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)** commented

[Martin Kanters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=martinkanters) Oh great, didn't realize that there is already a PR! Thanks!

**Comment by jira-importer** [comment]
**[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders)** commented

Fixed in [c3cf2943](https://github.com/apache/maven/commit/c3cf29438e3d65d6ee5c5726f8611af99d9a649a) with integration test in [9be9c249](https://github.com/apache/maven-integration-testing/commit/9be9c24999bfca177efdc5881b149a487ef5d6fc).
